tianshou/policy/modelfree/iqn.py

```diff
@@ -0,0 +1,103 @@
+import torch
+import numpy as np
+import torch.nn.functional as F
+from typing import Any, Dict, Optional, Union
+
+from tianshou.policy import QRDQNPolicy
+from tianshou.data import Batch, to_numpy
+
+
+class IQNPolicy(QRDQNPolicy):
+    """Implementation of Implicit Quantile Network. arXiv:1806.06923.
+
+    :param torch.nn.Module model: a model following the rules in
+        :class:`~tianshou.policy.BasePolicy`. (s -> logits)
+    :param torch.optim.Optimizer optim: a torch.optim for optimizing the model.
+    :param float discount_factor: in [0, 1].
+    :param int sample_size: the number of samples for policy evaluation.
+        Default to 32.
+    :param int online_sample_size: the number of samples for online model
+        in training. Default to 8.
+    :param int target_sample_size: the number of samples for target model
+        in training. Default to 8.
+    :param int estimation_step: the number of steps to look ahead. Default to 1.
+    :param int target_update_freq: the target network update frequency (0 if
+        you do not use the target network).
+    :param bool reward_normalization: normalize the reward to Normal(0, 1).
+        Default to False.
+
+    .. seealso::
+
+        Please refer to :class:`~tianshou.policy.QRDQNPolicy` for more detailed
+        explanation.
+    """
+
+    def __init__(
+        self,
+        model: torch.nn.Module,
+        optim: torch.optim.Optimizer,
+        discount_factor: float = 0.99,
+        sample_size: int = 32,
+        online_sample_size: int = 8,
+        target_sample_size: int = 8,
+        estimation_step: int = 1,
+        target_update_freq: int = 0,
+        reward_normalization: bool = False,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(model, optim, discount_factor, sample_size, estimation_step,
+                         target_update_freq, reward_normalization, **kwargs)
+        assert sample_size > 1, "sample_size should be greater than 1"
+        assert online_sample_size > 1, "online_sample_size should be greater than 1"
+        assert target_sample_size > 1, "target_sample_size should be greater than 1"
+        self._sample_size = sample_size  # for policy eval
+        self._online_sample_size = online_sample_size
+        self._target_sample_size = target_sample_size
+
+    def forward(
+        self,
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        model: str = "model",
+        input: str = "obs",
+        **kwargs: Any,
+    ) -> Batch:
+        model = getattr(self, model)
+        obs = batch[input]
+        obs_ = obs.obs if hasattr(obs, "obs") else obs
+        if model == "model_old":
```

Comment:
this is a bug. model now is a ``nn.Module` instead of str

Reply:
Good catch! Thanks!